In [1]:
import pandas as pd
import numpy as np

# DATA CLEANING

### <b>Ideas:<b>
    - Only look at people who were successfully iterviewed all 3 waves
    - Merge Wave1, Wave2 and Wave3 in order to look at the same population
    -
    -

    
    
    

In [32]:
Wave1 = pd.read_stata("Wave1.dta")
# display(Wave1.head(5))
# Wave1.info()

num_rows_start = Wave1.shape[0]

Wave1 = Wave1[Wave1['w1_nc_outcome'] == "Successfully Interviewed"]
deleted_rows = num_rows_start - Wave1.shape[0]

# display(Wave1.head(5))
# Wave1.info()
print("Number of people who were not interviewed in Wave 1: ", deleted_rows)

Number of people who were not interviewed in Wave 1:  0


In [33]:
Wave2 = pd.read_stata("Wave2.dta")
# display(Wave2.head(5))
# Wave2.info()

num_rows_start = Wave2.shape[0]

Wave2 = Wave2[Wave2['w2_nc_outcome'] == "Successfully Interviewed"]
deleted_rows = num_rows_start - Wave2.shape[0]

# display(Wave2.head(5))
# Wave2.info()
print("Number of people who were not interviewed in Wave 2: ", deleted_rows)

Number of people who were not interviewed in Wave 2:  1375


In [42]:
Wave3 = pd.read_stata("Wave3.dta")
# display(Wave3.head(5))
# Wave3.info()

num_rows_start = Wave3.shape[0]

Wave3 = Wave3[Wave3['w3_nc_outcome'] == "Successfully Interviewed"]
deleted_rows = num_rows_start - Wave3.shape[0]

# display(Wave3.head(5))
# Wave3.info()
print("Number of people who were not interviewed in Wave 3: ", deleted_rows)

Number of people who were not interviewed in Wave 3:  1968


We have now narrrowed down our dataset to only the people who were successfully interviewed.

In [37]:
frames = [Wave1, Wave2, Wave3]
temp = pd.merge(Wave1, Wave2, how='inner', on='pid')
df = pd.merge(temp, Wave3, how='inner', on='pid')

display(df.head(5))
print("Number of people who were successfully interviewed in all 3 waves: ", df.shape[0])

pid             w1_nc_outcome w1_nc_intrv_c w1_nc_intrv_d w1_nc_intrv_m  \
0  301067  Successfully Interviewed           609          15th           May   
1  301079  Successfully Interviewed           609          22nd           May   
2  301082  Successfully Interviewed           624           4th          June   
3  301086  Successfully Interviewed           644           4th          June   
4  301088  Successfully Interviewed           630          27th           May   

  w1_nc_intrv_y w1_nc_duration w1_nc_dob_m w1_nc_dob_y w1_nc_gen  ...  \
0          2020           15.2        July      1971.0     Woman  ...   
1          2020          15.37    February      1970.0     Woman  ...   
2          2020          25.27        June      1989.0       Man  ...   
3          2020          17.18        June      1972.0     Woman  ...   
4          2020          21.13         May      1978.0       Man  ...   

  w3_nc_smkprice_ld w3_nc_smkbrand_ld w3_nc_smkbrand_ld_o w3_nc_smkcigday  \
0               NaN               NaN                                 NaN   
1               NaN               NaN                                 NaN   
2               NaN               NaN                                 NaN   
3               NaN               NaN                                 NaN   
4               NaN               NaN                                 NaN   

  w3_nc_smkbrand_after w3_nc_smkbrand_after_o w3_nc_intres w3_nc_intlng1  \
0                  NaN                            Friendly        Sepedi   
1                  NaN                            Friendly        Sepedi   
2                  NaN                            Friendly      IsiXhosa   
3                  NaN                            Friendly       IsiZulu   
4                  NaN                             Hostile      IsiXhosa   

  w3_nc_intlng2 w3_nc_intlng3  
0           NaN           NaN  
1       Sesotho           NaN  
2           NaN           NaN  
3           NaN           NaN  
4       English           NaN  

[5 rows x 522 columns]

Number of people who were successfully interviewed in all 3 waves:  4508


In [46]:
personal_info_cols = [  'pid','w1_nc_dob_y', 'w2_nc_dob_y', 'w3_nc_dob_y',
                        'w1_nc_dob_m', 'w2_nc_dob_m', 'w3_nc_dob_m', 
                        'w1_nc_gen', 'w2_nc_gen', 'w3_nc_gen',
                        'w1_nc_popgrp',  'w3_nc_popgrp']

df_personal = df[personal_info_cols].copy()
df_personal

pid w1_nc_dob_y w2_nc_dob_y w3_nc_dob_y w1_nc_dob_m w2_nc_dob_m  \
0     301067      1971.0      1971.0      1971.0        July        July   
1     301079      1970.0      1970.0      1970.0    February    February   
2     301082      1989.0      1989.0      1989.0        June        June   
3     301086      1972.0      1972.0      1972.0        June        June   
4     301088      1978.0      1978.0      1978.0         May         May   
...      ...         ...         ...         ...         ...         ...   
4503  799572      1989.0      1990.0      1989.0      August      August   
4504  799619      1978.0      1978.0      1978.0   September     October   
4505  799644      1976.0      1976.0      1976.0     January     January   
4506  799780      1996.0      1996.0      1996.0        July        July   
4507  799922      1971.0      1971.0      1971.0     January     January   

     w3_nc_dob_m w1_nc_gen w2_nc_gen w3_nc_gen   w1_nc_popgrp   w3_nc_popgrp  
0           July     Woman     Woman     Woman  African/Black  African/Black  
1       February     Woman     Woman     Woman  African/Black  African/Black  
2           June       Man       Man       Man  African/Black  African/Black  
3           June     Woman     Woman     Woman  African/Black  African/Black  
4            May       Man       Man       Man  African/Black  African/Black  
...          ...       ...       ...       ...            ...            ...  
4503      August     Woman     Woman     Woman  African/Black  African/Black  
4504   September       Man       Man       Man  African/Black  African/Black  
4505     January     Woman     Woman     Woman  African/Black  African/Black  
4506        July     Woman     Woman     Woman       Coloured       Coloured  
4507     January       Man       Man       Man       Coloured       Coloured  

[4508 rows x 12 columns]

There are several cases where the values are inconsistent across the different waves. I believe this is an input error since many of the entires will match. For example, the year of birth in record #4503 is (1989, 1990, 1989) yet the month of birth and gender is consistent (August, Woman). 
To clean this up, we should look at the triplet of columns (that we are interested in - birth date, gender, etc) and set all 3 values to whichever one appears most often.

In [47]:
for index, row in df_personal.iterrows():
    years = row[['w1_nc_dob_y', 'w2_nc_dob_y', 'w3_nc_dob_y']]
    months = row[['w1_nc_dob_m', 'w2_nc_dob_m', 'w3_nc_dob_m']]
    genders = row[['w1_nc_gen', 'w2_nc_gen', 'w3_nc_gen']]
    races = row[['w1_nc_popgrp',  'w3_nc_popgrp']]


    YearOfBirth = years.value_counts().idxmax()
    MonthOfBirth = months.value_counts().idxmax()
    Gender = genders.value_counts().idxmax()   
    Race = races.value_counts().idxmax()

    

    df_personal.loc[index, 'YearOfBirth'] = YearOfBirth
    df_personal.loc[index, 'MonthOfBirth'] = MonthOfBirth
    df_personal.loc[index, 'Gender'] = Gender
    df_personal.loc[index, 'Race'] = Race

In [48]:
df_personal = df_personal.drop(personal_info_cols, axis=1)
df_personal

YearOfBirth MonthOfBirth Gender           Race
0         1971.0         July  Woman  African/Black
1         1970.0     February  Woman  African/Black
2         1989.0         June    Man  African/Black
3         1972.0         June  Woman  African/Black
4         1978.0          May    Man  African/Black
...          ...          ...    ...            ...
4503      1989.0       August  Woman  African/Black
4504      1978.0    September    Man  African/Black
4505      1976.0      January  Woman  African/Black
4506      1996.0         July  Woman       Coloured
4507      1971.0      January    Man       Coloured

[4508 rows x 4 columns]

In [49]:
df_dropped = df.drop(personal_info_cols, axis=1)
df_main = pd.concat([df_personal, df_dropped], axis=1)
df_main

YearOfBirth MonthOfBirth Gender           Race             w1_nc_outcome  \
0         1971.0         July  Woman  African/Black  Successfully Interviewed   
1         1970.0     February  Woman  African/Black  Successfully Interviewed   
2         1989.0         June    Man  African/Black  Successfully Interviewed   
3         1972.0         June  Woman  African/Black  Successfully Interviewed   
4         1978.0          May    Man  African/Black  Successfully Interviewed   
...          ...          ...    ...            ...                       ...   
4503      1989.0       August  Woman  African/Black  Successfully Interviewed   
4504      1978.0    September    Man  African/Black  Successfully Interviewed   
4505      1976.0      January  Woman  African/Black  Successfully Interviewed   
4506      1996.0         July  Woman       Coloured  Successfully Interviewed   
4507      1971.0      January    Man       Coloured  Successfully Interviewed   

     w1_nc_intrv_c w1_nc_intrv_d w1_nc_intrv_m w1_nc_intrv_y w1_nc_duration  \
0              609          15th           May          2020           15.2   
1              609          22nd           May          2020          15.37   
2              624           4th          June          2020          25.27   
3              644           4th          June          2020          17.18   
4              630          27th           May          2020          21.13   
...            ...           ...           ...           ...            ...   
4503           609          15th           May          2020          16.92   
4504           616          21st           May          2020           18.9   
4505           633          16th          June          2020          23.37   
4506           614          10th          June          2020           16.5   
4507           618          21st           May          2020          17.93   

      ... w3_nc_smkprice_ld      w3_nc_smkbrand_ld w3_nc_smkbrand_ld_o  \
0     ...               NaN                    NaN                       
1     ...               NaN                    NaN                       
2     ...               NaN                    NaN                       
3     ...               NaN                    NaN                       
4     ...               NaN                    NaN                       
...   ...               ...                    ...                 ...   
4503  ...               NaN                    NaN                       
4504  ...               5.0  Rudland & George (RG)                       
4505  ...               NaN                    NaN                       
4506  ...              80.0               Rothmans                       
4507  ...               NaN                    NaN                       

     w3_nc_smkcigday w3_nc_smkbrand_after w3_nc_smkbrand_after_o  \
0                NaN                  NaN                          
1                NaN                  NaN                          
2                NaN                  NaN                          
3                NaN                  NaN                          
4                NaN                  NaN                          
...              ...                  ...                    ...   
4503             NaN                  NaN                          
4504            10.0            Pall Mall                          
4505             NaN                  NaN                          
4506             8.0             Rothmans                          
4507             NaN                  NaN                          

                      w3_nc_intres w3_nc_intlng1 w3_nc_intlng2 w3_nc_intlng3  
0                         Friendly        Sepedi           NaN           NaN  
1                         Friendly        Sepedi       Sesotho           NaN  
2                         Friendly      IsiXhosa           NaN           NaN  
3                         Friendly       IsiZulu           Na

Above we can see the successful deletion of the duplicated DOB, gender and race columns, and the addition of the single columns with the correct values.